File: Orimath/Plugins/PluginExecutor.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Windows;
using Orimath.IO;
using Orimath.Reflection;
using Sssl;

namespace Orimath.Plugins
{
    internal static class PluginExecutor
    {
        public static Type[] LoadedPluginTypes { get; private set; } = Type.EmptyTypes;

        public static Type[] LoadedViewPluginTypes { get; private set; } = Type.EmptyTypes;

        public static PluginSetting Setting { get; private set; } = new PluginSetting();

        public static List<IConfigurablePlugin> ConfigurablePlugins { get; private set; } = new List<IConfigurablePlugin>();

        private static Type[] LoadPluginTypes()
        {
            var pluginDirectory = Path.Combine(Path.GetDirectoryName(Assembly.GetEntryAssembly()!.Location)!, "Plugins");

            if (Directory.Exists(pluginDirectory))
            {
                return Directory.GetFiles(pluginDirectory, "*.dll", SearchOption.TopDirectoryOnly)
                    .Select(Assembly.LoadFrom)
                    .SelectMany(asm => asm.GetExportedTypes())
                    .ToArray();
            }
            else
            {
                return Type.EmptyTypes;
            }
        }

        private static PluginSetting LoadSetting(Type[] types)
        {
            Setting = Settings.Load<PluginSetting>(SettingName.Plugin)!;
            if (Setting is null)
            {
                Setting = new PluginSetting
                {
                    PluginOrder = GetFullNames<IPlugin>(types)
                        .Concat(GetFullNames<IViewPlugin>(types))
                        .ToArray(),
                };
                SaveSetting();
            }
            
            return Setting;
        }

        public static void SaveSetting()
        {
            Settings.Save(SettingName.Plugin, Setting);
        }

        private static string[] GetFullNames<T>(Type[] types)
              where T : class
        {
            return types
                    .Where(t => t.IsClass && !t.IsAbstract && typeof(T).IsAssignableFrom(t))
                    .Select(t => t.FullName!)
                    .ToArray();
        }

        private static void ExecuteCore(PluginSetting setting, PluginArgs args, ViewPluginArgs viewArgs)
        {
            var pluginTypes = LoadedPluginTypes.ToDictionary(t => t.FullName);
            var viewPluginTypes = LoadedViewPluginTypes.ToDictionary(t => t.FullName);

            foreach (var fullName in setting.PluginOrder)
            {
                if (pluginTypes.TryGetValue(fullName, out var type) &&
                    Activator.CreateInstance(type) is IPlugin plugin)
                {
                    SetSetting(plugin, fullName, setting);
                    plugin.Execute(args);
                }

                if (viewPluginTypes.TryGetValue(fullName, out type) &&
                    Activator.CreateInstance(type) is IViewPlugin viewPlugin)
                {
                    SetSetting(viewPlugin, fullName, setting);
                    viewPlugin.Execute(viewArgs);
                }
            }
        }

        private static void SetSetting(object plugin, string fullName, PluginSetting setting)
        {
            if (plugin is IConfigurablePlugin configurable)
            {
                ConfigurablePlugins.Add(configurable);

                object? targetSetting;
                if (!setting.Settings.TryGetValue(fullName, out var sssl) ||
                    !sssl!.TryConvertTo(configurable.SettingType, out targetSetting))
                {
                    try
                    {
                        targetSetting = FastActivator.CreateInstance(configurable.SettingType);
                        setting.Settings[fullName] = SsslObject.ConvertFrom(targetSetting);
                    }
                    catch
                    {
                        targetSetting = null;
                    }

                }

                if (targetSetting is { })
                    configurable.Setting = targetSetting;
                else if (configurable.Setting is { })
                    setting.Settings[fullName] = SsslObject.ConvertFrom(configurable.Setting);
            }
        }

        private static IEnumerable<(Type, ViewAttribute)> GetViewTypes(Type[] types)
        {
            return types
                .Where(t => t.IsClass && !t.IsAbstract && typeof(FrameworkElement).IsAssignableFrom(t))
                .Select(t => (t, t.GetCustomAttribute<ViewAttribute>()!))
                .Where(tuple => tuple.Item2 is { });
        }

        public static IEnumerable<(Type, ViewAttribute)> ExecutePlugins(ViewPluginArgs viewArgs)
        {
            var args = new PluginArgs(viewArgs.Workspace);
            var types = LoadPluginTypes();
            var setting = LoadSetting(types);
            LoadedPluginTypes = types.Where(t => t.IsClass && !t.IsAbstract && typeof(IPlugin).IsAssignableFrom(t)).ToArray();
            LoadedViewPluginTypes = types.Where(t => t.IsClass && !t.IsAbstract && typeof(IViewPlugin).IsAssignableFrom(t)).ToArray();
            ExecuteCore(setting, args, viewArgs);

            return GetViewTypes(types);
        }
    }
}

File: Orimath/Plugins/PluginExecutor.fs
module Orimath.Plugins.PluginExecutor
open System
open System.Collections.Generic
open System.IO
open System.Reflection
open Orimath
open Orimath.IO
open SsslFSharp
open Orimath.Internal

let mutable loadedPluginTypes = Type.EmptyTypes

let mutable loadedViewPluginTypes = Type.EmptyTypes

let mutable setting = PluginSetting.CreateDefault()

let mutable configurablePlugins = List<IConfigurablePlugin>()

let private loadPluginTypes() =
    let pluginDirectory =
        Path.Combine(Path.GetDirectoryName(Assembly.GetEntryAssembly().Location), "Plugins")
    if Directory.Exists(pluginDirectory) then
        [|
            for file in Directory.GetFiles(pluginDirectory, "*.dll", SearchOption.TopDirectoryOnly) do
            for t in Assembly.LoadFrom(file).GetExportedTypes() do
            if t.IsClass && not t.IsAbstract then t
        |]
    else
        Type.EmptyTypes

let saveSetting() = Settings.save Settings.Plugin setting

[<RequiresExplicitTypeArguments>]
let getFullNames<'T>(types: Type[]) = [|
        for t in types do
        if typeof<'T>.IsAssignableFrom(t) then t.FullName
    |]

let private loadSetting (types: Type[]) =
    match Settings.load Settings.Plugin with
    | Some(s) when isNotNull (box s) ->
        setting <- s
        if isNull setting.PluginOrder then setting.PluginOrder <- Array.Empty()
        if isNull setting.Settings then setting.Settings <- Dictionary()
    | _ ->
        setting <- PluginSetting.CreateDefault()
        setting.PluginOrder <-
            Array.append (getFullNames<IPlugin> types) (getFullNames<IViewPlugin> types)
        saveSetting()
    setting

let private setSetting (plugin: obj) fullName (setting: PluginSetting) =
    match plugin with
    | :? IConfigurablePlugin as configurable ->
        configurablePlugins.Add(configurable)
        setting.Settings.TryGetValue(fullName)
        |> BoolOption.toOption
        |> Option.bind (Sssl.tryConvertToObj configurable.SettingType)
        |> Option.filter (isNull >> not)
        |> function
        | Some(targetSetting) -> configurable.Setting <- targetSetting
        | None ->
            if isNull configurable.Setting then
                configurable.Setting <- createInstance configurable.SettingType
            setting.Settings[fullName] <-
                Sssl.convertFromObj configurable.SettingType configurable.Setting
    | _ -> ()

let private executeCore (setting: PluginSetting) args viewArgs =
    let pluginTypes = Map.ofSeq (seq { for t in loadedPluginTypes -> t.FullName, t })
    let viewPluginTypes = Map.ofSeq (seq { for t in loadedViewPluginTypes -> t.FullName, t })
    for fullName in setting.PluginOrder do
        Map.tryFind fullName pluginTypes
        |> Option.bind createInstanceAs<IPlugin>
        |> Option.iter (fun plugin ->
            setSetting plugin fullName setting
            plugin.Execute(args))
        Map.tryFind fullName viewPluginTypes
        |> Option.bind createInstanceAs<IViewPlugin>
        |> Option.iter (fun viewPlugin ->
            setSetting viewPlugin fullName setting
            viewPlugin.Execute(viewArgs))

let ExecutePlugins (viewArgs: ViewPluginArgs) =
    let args = { Workspace = viewArgs.Workspace; FileManager = viewArgs.FileManager }
    let types = loadPluginTypes()
    let setting = loadSetting types
    loadedPluginTypes <- types |> Array.filter typeof<IPlugin>.IsAssignableFrom
    loadedViewPluginTypes <- types |> Array.filter typeof<IViewPlugin>.IsAssignableFrom
    executeCore setting args viewArgs
